akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -43,17 +43,37 @@ private[akka] final case class Filter[T](p: T ⇒ Boolean, decider: Supervision.
 }
 
 /**
- * INTERNAL API
- */
-private[akka] final case class TakeWhile[T](p: T ⇒ Boolean, decider: Supervision.Decider) extends PushStage[T, T] {
+  * INTERNAL API
+  */
+private[akka] final case class TakeWhile[T](p: T ⇒ Boolean) extends SimpleLinearGraphStage[T] {
+  override def initialAttributes: Attributes = DefaultAttributes.takeWhile
 
-  override def onPush(elem: T, ctx: Context[T]): SyncDirective =
-    if (p(elem))
-      ctx.push(elem)
-    else
-      ctx.finish()
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with OutHandler with InHandler {
+      override def toString = s"TakeWhile"
 
-  override def decide(t: Throwable): Supervision.Directive = decider(t)
+      val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
```

Comment:
Since it's only used when there is an actual exception, I think it makes sense to be a def.